src/uu/touch/src/touch.rs

```diff
@@ -236,96 +261,128 @@ pub fn uu_app() -> Command {
-                USimpleError::new(1, "Could not process the reference access time")
-            })?;
-            let mtime = filetime_to_datetime(&mtime).ok_or_else(|| {
-                USimpleError::new(1, "Could not process the reference modification time")
-            })?;
-            Ok((parse_date(atime, date)?, parse_date(mtime, date)?))
-        }
-        (Some(reference), None) => {
-            stat(Path::new(&reference), !matches.get_flag(options::NO_DEREF))
+/// Determine the access and modification time from a reference file
+fn determine_times_from_ref(
+    reference: &Path,
+    date: Option<&String>,
+    no_deref: bool,
+) -> UResult<(FileTime, FileTime)> {
+    if let Some(date) = date {
+        let (atime, mtime) = stat(reference, !no_deref)?;
+        let atime = filetime_to_datetime(&atime)
+            .ok_or_else(|| USimpleError::new(1, "Could not process the reference access time"))?;
+        let mtime = filetime_to_datetime(&mtime).ok_or_else(|| {
+            USimpleError::new(1, "Could not process the reference modification time")
+        })?;
+        Ok((parse_date(atime, date)?, parse_date(mtime, date)?))
+    } else {
+        stat(reference, !no_deref)
+    }
+}
+
+/// Execute the touch command.
+///
+/// # Errors
+///
+/// If the user doesn't have permission to access the file, or if one of the directory
+/// components of the file path doesn't exist
+pub fn touch(file: &InputFile, opts: &Options) -> UResult<()> {
+    // FIXME: find a way to avoid having to clone the path
+    // TODO (ysthakur) ^ Fixed with the Cow?
+    let (path, is_stdout) = match file {
+        InputFile::Stdout => (Cow::Owned(pathbuf_from_stdout()?), true),
+        InputFile::Path(path) => (Cow::Borrowed(path), false),
+    };
+    let path = path.as_path();
+
+    let metadata_result = if opts.no_deref {
+        path.symlink_metadata()
+    } else {
+        path.metadata()
+    };
+
+    if let Err(e) = metadata_result {
+        if e.kind() != std::io::ErrorKind::NotFound {
+            return Err(e.map_err_context(|| format!("setting times of {}", path.quote())));
         }
-        (None, Some(date)) => {
-            let timestamp = parse_date(Local::now(), date)?;
-            Ok((timestamp, timestamp))
+
+        if opts.no_create {
+            return Ok(());
         }
-        (None, None) => {
-            let timestamp = if let Some(ts) = matches.get_one::<String>(options::sources::TIMESTAMP)
-            {
-                parse_timestamp(ts)?
-            } else {
-                datetime_to_filetime(&Local::now())
-            };
-            Ok((timestamp, timestamp))
+
+        if opts.no_deref {
+            show!(USimpleError::new(
+                1,
+                format!(
+                    "setting times of {}: No such file or directory",
+                    path.quote()
+                )
+            ));
+            return Ok(());
```

Comment:
There's some more thinking to be done here. Basically, we should make the errors such that nushell can decide what to do on the error. I think we shouldn't show `USimpleError`s but instead make a proper error enum that we can expose to `nushell`. They can then match on it. At least, I'm interested in seeing what that would look like in the code. It's probably best to do it in a separate commit so we can compare with this implementation. We could also improve the API in another PR later.